KIPs/kip-81.md

```diff
@@ -0,0 +1,783 @@
+        if (choice == VoteChoice.Yes) { p.totalYes += votes; }
+        if (choice == VoteChoice.No) { p.totalNo += votes; }
+        if (choice == VoteChoice.Abstain) { p.totalAbstain += votes; }
+        if (!p.receipts[nodeId].hasVoted) { p.voters.push(nodeId); }
+    }
+
+    function checkQuorum(uint256 proposalId) external view returns(bool) {
+        Proposal storage p = proposals[proposalId];
+
+        ( , , , uint256 totalVotes, uint256 eligibleNodes) =
+            stakingTracker.getTrackerSummary(p.trackerId);
+
+        bool countQuorum = (p.voters.length >= (eligibleNodes / 3));
+        bool powerQuorum = ((p.totalYes + p.totalNo + p.totalAbstain) > (totalVotes / 3));
+        bool majorityYes = (p.totalYes > (p.totalNo + p.totalAbstain));
+        return ((countQuorum || powerQuorum) && majorityYes);
+    }
+}
+```
+
+#### Determining state
+
+The state of a proposal can be determined from stored variables, vote counts and current block number. Below is an example implementation of `state()` function.
+
+```solidity
+abstract contract Voting {
+    struct Proposal {
+        uint256 voteStart;      // propose()d block + votingDelay
+        uint256 voteEnd;        // voteStart + votingPeriod
+        uint256 queueDeadline;  // voteEnd + queueTimeout
+        uint256 eta;            // queue()d block + execDelay
+        uint256 execDeadline;   // queue()d block + execDelay + execTimeout
+        bool canceled;          // true if successfully cancel()ed
+        bool queued;            // true if successfully queue()d
+        bool executed;          // true if successfully execute()d
+    }
+    mapping(uint256 => Proposal) proposals;
+
+    function state(uint256 proposalId) external view returns (ProposalState) {
+        Proposal storage p = proposals[proposalId];
+
+        if (p.executed) {
+            return ProposalState.Executed;
+        } else if (p.canceled) {
+            return ProposalState.Canceled;
+        } else if (block.number < p.voteStart) {
+            return ProposalState.Pending;
+        } else if (block.number <= p.voteEnd) {
+            return ProposalState.Active;
+        } else if (!checkQuorum(proposalId)) {
+            return ProposalState.Failed;
+        }
+
+        if (!p.queued) {
+            if (block.number <= p.queueDeadline) {
+                return ProposalState.Passed;
+            } else {
+                return ProposalState.Expired;
+            }
+        } else {
+            if (block.number <= p.execDeadline) {
+                return ProposalState.Queued;
+            } else {
+                return ProposalState.Expired;
+            }
+        }
+    }
+}
+```
+
+#### Executing transactions
+
+The transactions in a passed proposal are executed in a order they appear in the `propose()` function arguments.  Each transaction is sent to `targets[i]`, carrying `values[i]` of KLAYs, with `calldatas[i]` as input data. Below is an example of `execute()` function.
```

Comment:
hmm can we check the data integrity? For example if the data is somewhat wrong, it could not be executed as we expected. How can we prevent the situtation?

Reply:
Perhaps the secretariat can try `klay_call("Voting.execute(id)")` to see if it reverts. Though the try must be after execDelay.